proposals/XXXX-typed-based-private.md

```diff
@@ -0,0 +1,146 @@
+        self.name = name
+        self.age = age
+    }
+
+    func greet() {
+        // age is accessible because it is defined in the same declaration
+        // secretAge is not because it is defined in a nested type
+        // Therefore, following piece of code would generate a compilation error:
+        // if age < gender.secrecyAge {
+        
+        // Instead:
+        if Gender.shouldRevealAge(self) {
+            // fullGreeting is accessible because it is defined in an extension
+            // of the same type in the same file
+            print(fullGreeting)
+        } else {
+            print(greeting)
+        }
+    }
+}
+
+// private at the top-level scope continues to be equivalent to fileprivate
+private extension Person {
+    // fullGreeting is implictly private due to the extension's modifier
+    var fullGreeting: String {
+        // age is accessible because it is defined in the declaration of the extension's type in the
+        // same file
+        return "Hello, my name is \(name), I'm \(age) years old."
+    }
+}
+
+extension Person {
+    enum Gender {
+        case male
+        case female
+
+        private var secrecyAge: Int {
+            switch self {
+            case .male: return 60
+            case .female: return 50
+            }
+        }
+
+        static func shouldRevealAge(_ person: Person) -> Bool {
+            // age is accessible because it was declared in the same file and we are
+            // in the declaration of a type nested in the type that declared age
+            return person.age < person.gender.secrecyAge
+        }
+    }
+}
+
+extension Gender {
+    static func leakAge(of person: Person) {
+        // age is accessible because we are in the extension of a type nested in the
+        // type same file
+        return person.age
+    }
+}
+
+class VIP: Person {
+   var leakedAge: Int {
+       // error: age is not visible because it is defined in another type
+       return age
+   }
+}
+```
+
+### Other.swift
+
+```swift
+extension Person : CustomStringConvertible {
+    var desription: String {
+        return fullGreeting
+        // error: fullGreeting is not available because it is defined in another file
+    }
+}
+```
+
+## Source compatibility
+
+In Swift 3 compatibility mode, the compiler will continue to treat `private` as before. In Swift 4 mode, the compiler will modify the semantics of `private` to follow the rules of this proposal. No migration will be necessary as this proposal merely broadens the visibility of `private`.
+
+Cases where a type had `private` declarations with the same signature in the same type/extension but in different scopes will produce a compiler error in Swift 4. For example, the following piece of code compiles in Swift 3 compatibilty mode but generates a `Invalid redeclaration of 'bar()'` error in Swift 4 mode:
+
+```swift
+struct Foo {
+    private func bar() {}
+}
+
+extension Foo {
+    private func bar() {}
```

Comment:
I meant if an explicit private method in a extension could be accessed in another extension?